app/src/main/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepository.java

```diff
@@ -308,6 +338,61 @@ public void saveCategories(List<Category> categories) {
         }
     }
 
+    /**
+     * Delete rows from Ingredient, IngredientName and IngredientsRelation
+     * set the autoincrement to 0
+     */
+    @Override
+    public void deleteIngredientCascade(){
+        ingredientDao.deleteAll();
+        ingredientNameDao.deleteAll();
+        ingredientsRelationDao.deleteAll();
+        DaoSession daoSession = OFFApplication.getInstance().getDaoSession();
+        daoSession.getDatabase().execSQL("update sqlite_sequence set seq=0 where name in ('" + ingredientDao.getTablename() + "', '" + ingredientNameDao.getTablename() + "', '" + ingredientsRelationDao.getTablename() + "')");
+    }
+
+    /**
+     * Ingredients saving to local database
+     * <p>
+     * Ingredient and IngredientName has One-To-Many relationship, therefore we need to save them separately.
+     */
+    @Override
+    public void saveIngredients(List<Ingredient> ingredients) {
+        db.beginTransaction();
+        try {
+            for (Ingredient ingredient : ingredients) {
+                ingredientDao.insertOrReplace(ingredient);
+                for (IngredientName ingredientName : ingredient.getNames()) {
+                    ingredientNameDao.insertOrReplace(ingredientName);
+                }
+                for (IngredientsRelation ingredientsRelation : ingredient.getParents()) {
+                    ingredientsRelationDao.insertOrReplace(ingredientsRelation);
+                }
+                for (IngredientsRelation ingredientsRelation : ingredient.getChildren()) {
+                    ingredientsRelationDao.insertOrReplace(ingredientsRelation);
+                }
+            }
+
+            db.setTransactionSuccessful();
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
Use Log

Reply:
I don't understand what do you mean by "Use Log".

Reply:
replace e.printStackTrace by using Log class Log.e(....,e)

Reply:
Well, the "catch" I've found in OFF code have an "e.printStackTrace".
Is it really important for you ?

Reply:
Yes i already see it ;) It's just a common practice to use a Log tool. Maybe this error is not really important and we should only log on debug mode to avoid filing logcat..